pkg/server/helpers.go

```diff
@@ -496,3 +499,89 @@ func getRuntimeOptions(c containers.Container) (interface{}, error) {
 	}
 	return opts, nil
 }
+
+func getCurrentOOMScoreAdj() int {
+	b, err := ioutil.ReadFile("/proc/self/oom_score_adj")
+	if err != nil {
+		return 0
+	}
+	i, err := strconv.Atoi(string(b))
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func restrictOOMScoreAdj(spec *runtimespec.Spec) {
+	currentOOMScoreAdj := getCurrentOOMScoreAdj()
+	if spec.Process.OOMScoreAdj != nil && *spec.Process.OOMScoreAdj < currentOOMScoreAdj {
+		*spec.Process.OOMScoreAdj = currentOOMScoreAdj
+	}
+}
+
+// parseAutoOrBool returns nil for "auto" and "", *bool for bool string.
+func parseAutoOrBool(v string) (*bool, error) {
+	b, err := strconv.ParseBool(v)
+	if err == nil {
+		return &b, nil
+	}
+	if v == "" || v == "auto" {
```

Comment:
> Suggest: using a string for a rootless option, "auto" should be default.

@mikebrow Let's not change the default behavior... It is dangerous to me.
You never know what people have been doing.

Reply:
if true we might need to change runc.. I think they used auto there?

Reply:
more thought is needed I think.

Reply:
Today's behavior is that we always enable cgroup, no restrict on OOMScore.

The default config should behave that way.

Reply:
Can we remove `auto` and revert the type to bool?

Reply:
> Can we remove auto and revert the type to bool?

I'm fine with either, as long as the default behavior is not affected.

Actually, I prefer only making the type bool, but error out with clear message if containerd is in user namespace, and unsupported options are enabled, e.g. cgroups, apparmor etc.

Reply:
that works too! and covers the reason I figured we needed auto mode.

Reply:
> Actually, I prefer only making the type bool, but error out with clear message if containerd is in user namespace, and unsupported options are enabled, e.g. cgroups, apparmor etc. Instead of silently trying to do something smart.

Hm, I found that in this way, the default config will be kept the same, which is good and clear.

However, the default behavior is still changed:
* Before: Containerd in user namespace won't return error directly when apparmor or cgroup is enabled;
* Now: Containerd in user namespace returns error in that case.

However, if containerd in user namespace doesn't support those features, even in the previous behavior we'll eventually get an error. So it seems fine to me.

Reply:
Updated to print a warning, not an error (because even in userns cgroups might be available especially when running in LXD or with pam_cgfs.so)